spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageResourceBucket.java

```diff
@@ -0,0 +1,158 @@
+/*
+ *  Copyright 2017 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.storage;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+
+import com.google.cloud.storage.Bucket;
+import com.google.cloud.storage.BucketInfo;
+import com.google.cloud.storage.Storage;
+
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.WritableResource;
+import org.springframework.util.Assert;
+
+/**
+ * Implements {@link WritableResource} for buckets in Google Cloud Storage.
+ */
+public class GoogleStorageResourceBucket implements WritableResource {
+
+	private final Storage storage;
+
+	private final String bucketName;
+
+	public GoogleStorageResourceBucket(Storage storage, String bucketName,
+			boolean createBucketIfNotExists) {
+		Assert.notNull(storage, "Storage object can not be null");
+		Assert.notNull(bucketName, "Bucket name can not be null");
+
+		this.storage = storage;
+		this.bucketName = bucketName;
+
+		createBucketIfRequired(createBucketIfNotExists);
```

Comment:
I actually meant not doing this in the constructor at all. Instead, the user can call `create()` if they want to create it explicitly, or we create it for them automatically in any appropriate methods (`createRelative()` is really the only one that comes to mind) if `createBucketIfNotExists` is `true`.

Reply:
I think it's just a difference of opinion. I prefer doing it in the constructor, and we control it with the auto-create boolean argument.

Reply:
It's more than just opinion.
The constructor should not contain extraneous logic that should belong in other methods. Later on this PR you're creating the file remotely in GCS by calling the constructor. I would not expect that as a Java developer.
There is no reason why the status of an object of this kind in memory should map to the status of the file in GCS. That's why methods like `exists()` exist.

It also makes unit testing objects of this kind a bummer.
That method should probably also throw an `IOException` in case the file couldn't be constructed, which is a code smell coming from a constructor.

Reply:
I really think your explanation still sounds just like your opinion. I can explain my perspective too:

- a user can create a GSRB setting the auto-create bool
- the user doesn't need to know if the bucket already exists: that's the whole point of the auto-create functionality. 
- the `exists` function works as expected

As a Java developer I think this is pretty simple and intuitive.

Reply:
In the case of a GCS object resource we only create it when attempting to write to it. I don't see why it should be any different for a bucket resource. The "write" operation for a Bucket is `.createRelative()`. So, I think that's where the call to `createBucketIfRequired` belongs.

`GoogleStorageResourceBucket`, just like `java.util.File` is simply a handle to a resource that may or may not exist. So, auto-creation in the constructor doesn't make sense to me. I would also make `createBucketIfRequired` public.

Reply:
Yea this makes sense. I'll make the change